tuplex/io/src/OrcTypes.cc

```diff
@@ -0,0 +1,136 @@
+    std::vector<python::Type> types;
+    for (uint64_t i = 0; i < rowType.getSubtypeCount(); i++) {
+        auto hasNull = columnHasNull.at(i);
+        auto el = orcTypeToTuplex(*rowType.getSubtype(i), hasNull);
+        if (hasNull) {
+            types.push_back(python::Type::makeOptionType(el));
+        } else {
+            types.push_back(el);
+        }
+    }
+    return python::Type::makeTupleType(types);
+}
+
+python::Type orcTypeToTuplex(const ::orc::Type &type, bool hasNull) {
+    switch (type.getKind()) {
+        case ::orc::BOOLEAN: {
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::BOOLEAN);
+            } else {
+                return python::Type::BOOLEAN;
+            }
+        }
+        case ::orc::BYTE:
+        case ::orc::SHORT:
+        case ::orc::INT:
+        case ::orc::LONG: {
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::I64);
+            } else {
+                return python::Type::I64;
+            }
+        }
+        case ::orc::FLOAT:
+        case ::orc::DOUBLE: {
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::F64);
+            } else {
+                return python::Type::F64;
+            }
+        }
+        case ::orc::VARCHAR:
+        case ::orc::CHAR:
+        case ::orc::STRING: {
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::STRING);
+            } else {
+                return python::Type::STRING;
+            }
+        }
+        case ::orc::LIST: {
+            auto elementType = orcTypeToTuplex(*type.getSubtype(0), hasNull);
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::makeListType(elementType));
+            } else {
+                return python::Type::makeListType(elementType);
+            }
+        }
+        case ::orc::MAP: {
+            auto keyType = orcTypeToTuplex(*type.getSubtype(0), hasNull);
+            auto valueType = orcTypeToTuplex(*type.getSubtype(1), hasNull);
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::makeDictionaryType(keyType, valueType));
+            } else {
+                return python::Type::makeDictionaryType(keyType, valueType);
+            }
+        }
+        case ::orc::STRUCT: {
+            std::vector<python::Type> types;
+            for (uint64_t i = 0; i < type.getSubtypeCount(); i++) {
+                auto el = orcTypeToTuplex(*type.getSubtype(i), hasNull);
+                types.push_back(el);
+            }
+            if (hasNull) {
+                return python::Type::makeOptionType(python::Type::makeTupleType(types));
+            } else {
+                return python::Type::makeTupleType(types);
+            }
+        }
+        default:
+            throw std::runtime_error("Orc row type unable to be converted to Tuplex type");
```

Comment:
can we maybe get a specific orc type to string message? This would help the user more to debug in case :)